File: src/leetcode/solved/linkedList/AddTwoNumbers.java
package leetcode.solved.linkedList;

public class AddTwoNumbers {
    //#2 Add Two Numbers
    static class ListNode{
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    public static void main(String[] args) {
        ListNode l1 = new ListNode(3);
        ListNode l2 = new ListNode(9);
        l1.next = new ListNode(7);
        l2.next = new ListNode(2);


        ListNode asd = addTwoNumbers(l1, l2);

        while(asd!=null){
            System.out.print("->" + asd.val);
            asd = asd.next;
        }
    }

    public static ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode resultList = new ListNode(0);
        ListNode iterative = resultList;

        int carry = 0;
        while (l1!=null && l2!=null){
            int x = (l1.val + l2.val + carry)%10;
            carry = (l1.val+l2.val+carry)/10;
            iterative.next = new ListNode(x);
            iterative = iterative.next;
            l1 = l1.next;
            l2 = l2.next;
        }

        ListNode rest = l1!=null ? l1 : l2;
        if(rest!=null || carry>0){
            while (carry>0 && rest!=null){
                int x = (rest.val + carry)%10;
                carry = (rest.val+carry)/10;
                rest = rest.next;
                iterative.next = new ListNode(x);
                iterative = iterative.next;
            }
            if(rest!=null)
                iterative.next = rest;
            else if(carry>0)
                iterative.next = new ListNode(carry);
        }
        resultList = resultList.next;

        return resultList;
    }

}

File: src/leetcode/solved/linkedList/AddTwoNumbers.kt
class ListNode(var `val`: Int) {
    var next: ListNode? = null
}

fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
    var res: ListNode? = null
    var iterator: ListNode? = res
    var carry = 0

    var node1 = l1
    var node2 = l2

    while (node1!=null || node2!=null){
        val sum: Int = (node1?.`val`?:0) + (node2?.`val`?:0) + carry
        carry = sum/10

        if(res == null){
            res = ListNode(sum%10)
            iterator = res
        } else{
            iterator?.next = ListNode(sum%10)
            iterator = iterator?.next
        }

        node1 = node1?.next
        node2 = node2?.next
    }

    if(carry>0)
        iterator?.next = ListNode(carry)

    return res
}

fun main(){

    var l1 = ListNode(2)
    var l2 = ListNode(4)
    var l3 = ListNode(3)

    l1.next = l2
    l2.next = l3

    var ll1 = ListNode(5)
    var ll2 = ListNode(6)
    var ll3 = ListNode(4)

    ll1.next = ll2
    ll2.next = ll3

    var asd: ListNode? = addTwoNumbers(l1, ll1)

    while (asd!=null){
        print(" ${asd.`val`}")
        asd = asd.next
    }
}